backend/corpora/dataset_processing/download.py

```diff
@@ -0,0 +1,157 @@
+import logging
+import threading
+import queue
+import requests
+
+from ..common.corpora_orm import DbDatasetProcessingStatus, UploadStatus
+from ..common.entities import Dataset
+from ..common.utils.db_utils import db_session_manager
+from ..common.utils.math_utils import MB
+
+logger = logging.getLogger(__name__)
+
+
+class ProgressTracker:
+    def __init__(self, file_size: int):
+        self.file_size: int = file_size
+        self._progress: int = 0
+        self.progress_lock: threading.Lock = threading.Lock()  # prevent concurrent access of ProgressTracker._progress
+        self.stop_updater: threading.Event = threading.Event()  # Stops the update_progress thread
+        self.stop_downloader: threading.Event = threading.Event()  # Stops the downloader threads
+        self.error: queue.Queue = queue.Queue()  # Track errors
```

Comment:
> but isn't this written once by a single thread and read once by another?

This was the impression I was under but since two threads are using the error queue so it's possible they could both fail.